### PR TITLE
optional utilities and monadic interface

### PR DESCRIPTION
**Pull request by xlauko**

### CODE REVIEW

File: core/include/gap/core/concepts.hpp [hunk 1]
```diff
@@ -66,6 +69,20 @@ namespace gap
         std::invoke(std::forward< F >(f), std::forward< Args >(args)...);
     };
 
+    namespace detail {
+        template< typename B >
+        concept boolean_testable = std::convertible_to< B, bool >;
```
Comment (lkorenc) [inline core/include/gap/core/concepts.hpp:74]: Unless this is used somewhere else I would inline it.

File: core/include/gap/core/optional.hpp [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        {}
+
+        template< typename U = value_type >
+        requires(std::is_constructible_v< base_type, U&& >)
+        explicit(!std::is_convertible_v< U&&, value_type >)
+        constexpr optional_wrapper( U &&other )
+            : base_type(std::forward< U >(other))
+        {}
+
+        base& unwrap() & { return *this; }
```
Comment (lkorenc) [inline core/include/gap/core/optional.hpp:150]: Why is this not `value()` from the parent class?

File: core/include/gap/core/optional.hpp [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        const base& unwrap() const& { return *this; }
+        base&& unwrap() && { return std::move(*this); }
+        const base&& unwrap() const&& { return std::move(*this); }
+
+        template< typename F, optional_like R = std::invoke_result_t< F, value_type& > >
+        constexpr auto and_then( F&& f ) & {
+            if (*this) {
```
Comment (lkorenc) [inline core/include/gap/core/optional.hpp:157]:
```suggestion
            if (this->has_value()) {
```
To improve readability?

File: core/include/gap/core/optional.hpp [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+namespace gap {
+
+    //
+    // optional concepts
+    //
+    template< typename T >
+    concept optional_like = requires (T a) {
+        { a.value() }     -> convertible_to< typename T::value_type >;
+        { a.has_value() } -> convertible_to< bool >;
+        { *a } -> convertible_to< typename T::value_type >;
```
Comment (lkorenc) [inline core/include/gap/core/optional.hpp:21]: What about default constructibility? Is it not needed for `and_then` and others?

File: core/include/gap/core/optional.hpp [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        constexpr auto and_then( F&& f ) & {
+            if (*this) {
+                return std::invoke(std::forward< F >(f), this->value());
+            } else {
+                return std::remove_cvref_t< R >();
+            }
+        }
+
+        template< typename F, optional_like R = std::invoke_result_t< F, const value_type& > >
+        constexpr auto and_then( F&& f ) const& {
```
Comment (lkorenc) [inline core/include/gap/core/optional.hpp:165]:
```suggestion
        constexpr R and_then( F&& f ) const& {
```
or maybe as trailing return? Another bit to help reader.

File: core/include/gap/core/optional.hpp [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+            if (*this) {
+                return std::invoke(std::forward< F >(f), this->value());
+            } else {
+                return std::remove_cvref_t< R >();
```
Comment (lkorenc) [inline core/include/gap/core/optional.hpp:169]: I am not sure why `std::remove_cvref_t` is needed, can you give me such example?

**Review by lkorenc — APPROVED** [review]

File: core/include/gap/core/optional.hpp (hunk 2 above)
Comment (xlauko) [reply core/include/gap/core/optional.hpp:150]: it is completely different thing :P `value` returns `**this` stored value, whereas `unwrap` casts to base class `*this`

File: core/include/gap/core/optional.hpp (hunk 3 above)
Comment (xlauko) [reply core/include/gap/core/optional.hpp:157]: generates overhead: https://godbolt.org/z/z31zc3d7v :D

File: core/include/gap/core/optional.hpp (hunk 4 above)
Comment (xlauko) [reply core/include/gap/core/optional.hpp:21]: true

File: core/include/gap/core/optional.hpp (hunk 6 above)
Comment (xlauko) [reply core/include/gap/core/optional.hpp:169]: what if `R` is `const &T` ?

File: core/include/gap/core/optional.hpp (hunk 5 above)
Comment (xlauko) [reply core/include/gap/core/optional.hpp:165]: it is not `R` (see the following answer)

**Pull request merged**